### ISSUE TITLE
Slice Timing

### ISSUE THREAD

**Issue by araikes** [BIDS-Apps/MRtrix3_connectome#59]
Hi @Lestropie ,

I got CUDA implemented to work with docker and wanted to test using eddy_cuda9.1. However, I'm running into a problem. At the point of dwipreproc, the slice timing in the header has been lost for dwipreproc_in. That image ends up having a single slice timing value, rather then e.g. 60 values.

I know this was suggested in #44. Thoughts on a solution?

**Comment by Lestropie** [comment]
> I got CUDA implemented to work with docker

Nice! Would be awesome to get a PR to bring in such changes.

> At the point of dwipreproc, the slice timing in the header has been lost for dwipreproc_in. That image ends up having a single slice timing value, rather then e.g. 60 values.

Can you confirm that valid slice timing information is present in both the `dwi/` and `fmap/` JSONs, but collapses to a single value as soon as the concatenation is performed, and is not attributed to some other step in between that concatenation and the `dwipreproc` step?

Given the concatenation is *not* along the slice encoding axis, the information should be preserved as long as the contents of the key-value entry are identical. Do the slice timing data differ at all between the input volumes?

I'll have to think a little more about that area of the pipeline. There has been at least some doubt raised about the validity of concatenating data with different phase encoding directions for running `dwidenoise`; I also need to better deal with the prospect that images in the `fmap/` directory may not have the same spatial resolution as those in the `dwi/` directory, in which case they cannot be concatenated without interpolation, and one should not interpolate values prior to `dwidenoise`.

Modifications to the *MRtrix3* Python API for the upcoming release do include a mechanism that can be used to better utilise JSON files for header key-value override. So once that release is done and I finally get to update this tool accordingly, that will be a more convenient mechanism for dealing with such issues.

If you simply want to get your version of the script running, perhaps the easiest solution would be to strip down that concatenation process to only operate across DWIs, i.e. don't combine & then separate out the `fmap` files, and instead just concatenate the latter separately, run through `mrdegibbs`, and feed to `dwipreproc -se_epi`.

**Comment by araikes** [comment]
Hi @Lestropie 

Regarding a PR for CUDA implementation, I can certainly do that but I before I do, I wanted to run past you what I had to do in order to get it working:

1. I had to install Docker CE 19.03 and nvidia-docker (https://github.com/NVIDIA/nvidia-docker). Older versions of nvidia-docker work with older versions of docker, but the current version allows for accessing GPUs directly.
1. I had to change the Ubuntu version in the Dockerfile to 16.04. This is for a couple of reasons. First eddy_cuda is updated to 9.1 as of FSL v. 6.0.1 and 9.1 is available for 16.04 but not 18.04. Second, we are running 16.04 here. I don't know if it'll make a difference. The major thing I can potentially think of is that the default python version is 2.7.2, not 3.5+.
1. Using nvidia-docker as a pass through, I grabbed the 9.1 CUDA files in the Dockerfile and then at the end set ENV variables to have access to the GPU cores.

Like I said, I can do a PR for it but it will also take some instructions and probably necessitates two versions of the Dockerfile (a CUDA version and a non-CUDA version).

Regarding the SliceTiming, I can confirm that it is a keyval in the original `dwi/` image, present after converting from nii.gz to .mif, and is the correct length. The `fmap/` file does not have slice timing. After concatenation, there is no 'SliceTiming' keyval. 

So I've tested it out not concatenating the fmap and dwi images and the slice timing is preserved. However, there's a new issue. The object for 'SliceTiming' is a list, for example:
```
[[4.64], [0.0], [4.795], [0.155], [4.95], [0.31], [5.105], [0.465], [5.26], [0.62], [5.4125], [0.7725], [5.5675], [0.9275], [5.7225], [1.0825], [5.8775], [1.2375], [6.0325], [1.3925], [6.1875], [1.5475], [6.3425], [1.7025], [6.495], [1.8575], [6.65], [2.01], [6.805], [2.165], [6.96], [2.32], [7.115], [2.475], [7.27], [2.63], [7.425], [2.785], [7.5775], [2.94], [7.7325], [3.0925], [7.8875], [3.2475], [8.0425], [3.4025], [8.1975], [3.5575], [8.3525], [3.7125], [8.5075], [3.8675], [8.6625], [4.0225], [8.815], [4.175], [8.97], [4.33], [9.125], [4.485]]
```
The length of that list is 60, same number as the number of slices in my image. However, dwipreproc sees it as length = 1 because line 230 of dwipreproc accesses only the first item `slice_timing = dwi_header.keyval()['SliceTiming'][0]` instead of `slice_timing = dwi_header.keyval()['SliceTiming']`. 

I don't know if this is a Python 2.7 vs Python 3.5+ thing but it disallows slice-to-volume correction at present.

**Comment by Lestropie** [comment]
> I had to change the Ubuntu version in the Dockerfile to 16.04.

Don't think that would be a massive option; I jumped from 14.04 to 18.04 simply because I could. 

> Additionally, Python 2.7 is the default for Ubuntu 16.04. 

Also shouldn't be a problem; we've made sure everything in *MRtrix3* works with both Python 2 and 3, and this script is no different.

> and probably necessitates two versions of the Dockerfile (a CUDA version and a non-CUDA version).

If there's a way to do that such that DockerHub can still identify and build both, that'd be awesome. Can look into that if you make a PR.

> The object for 'SliceTiming' is a list, for example:

Okay, that might be trickier to sort out. Acquiring those data relies on a `mrinfo -json_all` call, followed by a Python `json.load()`. From the code I can infer that when I wrote it, the Python JSON library was giving me those data as a list of lists; the outer list had only one entry, the contents of which was a list of floats. At your end, the same call is giving a list of lists, but the outer list has a number of entries corresponding to the number of slices, and each of those lists contains only one entry. So the question is, why are the data produced by the same code giving two different results:

- The Python JSON library does appear to have some degree of type identification, as I've never had to do any explicit type conversion in the `image.Header` constructor. This could conceivably vary slightly between Python versions.

- Anything else could conceivably result in different outcomes between Python versions; e.g. how string data are handled.

- The `mrinfo -json_all` call itself depends on "JSON for Modern C++" as an external library. A change there could conceivably change how the raw text data stored in the image header are exported to JSON.

Ideally I'd like to narrow this down. However if the data can be produced in this format in some instances, then the `dwipreproc` script probably needs to support both, which will require wrapping up some logic to disentangle the two and reformat either into a simple list of floats.

**Comment by araikes** [comment]
So a little poking and prodding later:

I ran the following:
```
# Set JSON Path
json_path = /path/to/original/dwi.json

# Load the JSON
with open(json_path, 'r') as f:
    json_elements = json.load(f)

# Save slice time element
slice_time = json_elements['SliceTiming']
 
# Print it
print(slice_time)

## Output
[4.64, 0, 4.795, 0.155, 4.95, 0.31, 5.105, 0.465, 5.26, 0.62, 5.4125, 0.7725, 5.5675, 0.9275, 5.7225, 1.0825, 5.8775, 1.2375, 6.0325, 1.3925, 6.1875, 1.5475, 6.3425, 1.7025, 6.495, 1.8575, 6.65, 2.01, 6.805, 2.165, 6.96, 2.32, 7.115, 2.475, 7.27, 2.63, 7.425, 2.785, 7.5775, 2.94, 7.7325, 3.0925, 7.8875, 3.2475, 8.0425, 3.4025, 8.1975, 3.5575, 8.3525, 3.7125, 8.5075, 3.8675, 8.6625, 4.0225, 8.815, 4.175, 8.97, 4.33, 9.125, 4.485]
```
So, as can be seen, `json.load` is yielding the list of floats as expected.

However, when the JSON is imported when converting from the original image to dwi1.mif (line [211 ](https://github.com/BIDS-Apps/MRtrix3_connectome/blob/12911ab81074c4582f5d58a8b2249be7cfe51435/mrtrix3_connectome.py#L211)), it appears that it's at this point that it converts to a list of lists.

```
# Convert from original to dwi1.mif
run.command('mrconvert ' + entry + grad_import_option + json_import_option +
            path.toTemp('dwi' + str(dwi_index) + '.mif', True))

# Get slice timing keyval
image.Header('dwi1.mif').keyval()['SliceTiming']

# Output
[[4.64], [0.0], [4.795], [0.155], [4.95], [0.31], [5.105], [0.465], [5.26], [0.62], [5.4125], [0.7725], [5.5675], [0.9275], [5.7225], [1.0825], [5.8775], [1.2375], [6.0325], [1.3925], [6.1875], [1.5475], [6.3425], [1.7025], [6.495], [1.8575], [6.65], [2.01], [6.805], [2.165], [6.96], [2.32], [7.115], [2.475], [7.27], [2.63], [7.425], [2.785], [7.5775], [2.94], [7.7325], [3.0925], [7.8875], [3.2475], [8.0425], [3.4025], [8.1975], [3.5575], [8.3525], [3.7125], [8.5075], [3.8675], [8.6625], [4.0225], [8.815], [4.175], [8.97], [4.33], [9.125], [4.485]]
```
Using the `-clear_property` and the `-append_property` options in `mrconvert`, I cleared the SliceTiming keyval and added the slice times from the list of floats I saved using `json.load()['SliceTiming`]. (BTW: I had to this one at a time in a loop. Saving the list of floats to an object and appending that object yielded an error where it thought I had 61 output files). This yields the same list of 1D lists.

So it seems that the CPP JSON method is converting it. I can see two possible methods for "easily" resolving the issue:

1. Save the slice timings using `json_load`, create a `--slspec` file, and pass that argument to eddy in dwipreproc.
2. Change the dwipreproc call to grab the length of the list of lists `len(dwi_header.keyval()['SliceTiming'])`. I just don't know if there are going to be downstream issues if it stays in list of lists format.

**Comment by Lestropie** [comment]
Ah. I think it just registered what I've done wrong. 

As soon as I first write the slice timing information from an input DICOM image to an image header ([in RAM](https://github.com/MRtrix3/mrtrix3/blob/3.0_RC3/core/file/dicom/mapper.cpp#L241), before it's even gone to disk), I use commas rather than spaces. The JSON exporter must therefore interpret that as a sort of newline, rather than turning the comma-separated list of values into a list of floats. 

So the issue is that all the JSONs I've worked with have been themselves erroneous due to having been generated by *MRtrix3*, and I've tweaked `dwipreproc` to import buggy data; you've now provided it with a JSON generated through something other than *MRtrix3*, and the `dwipreproc` fudge has been exposed. :man_facepalming: 

I'll need to:

1. Fix the DICOM import in *MRtrix3*

2. Add a fudge in `dwipreproc` to detect the erroneous JSON data and modify it accordingly (since someone could update the software but still try to run it on already converted data)

3. Import the change into this tool.

**Comment by araikes** [comment]
Ok, related to this topic:

I now have a Dockerfile and edits to the Python script in place that negates the need for multiple Dockerfiles in order to enable CUDA processing. It should also be flexible enough to use as a stand alone (non-Docker) script as it only adds one Python dependency (numba: https://numba.pydata.org/numba-doc/dev/index.html).

There will be some required documentation in terms of setting up Docker to allow CUDA and I don't know how this might work with Singularity but I can post a PR with the changes in the `dev` branch for you to review if you're interested.

**Comment by Lestropie** [comment]
Would definitely be interested in getting this working; I tend to use the script outside of a container myself purely to utilise CUDA. I'd likely need to see the code to understand why numba is required; ideally if it's only there for getting CUDA working in the container itd be made a non-essential dependency outside of the container with some `except ImportError` trickery. `dev` is a bit of a moving target at the moment given the incoming requests, but I can always assist should things diverge.

**Comment by araikes** [comment]
Doing this on my phone, so I'll be more verbose later.

In what I have set up, I'm exposing the GPUs to Docker using nvidia-docker installed locally (see initial post above). This adds a `--gpus` flag to the Docker call if I want CUDA capability. CUDA libs are then installed in the Dockerfile and exist regardless of whether CUDA is being used it not.

In the script, I'm using `len(numba cuda.gpus)`  in a `try except else` statement. If there are no CUDA GPUs, it throws an error and switches to `eddy_openmp`. Otherwise it uses CUDA. This way, you don't have to remove `eddy_cuda` in the Dockerfile because it only matters if there CUDA cores available.

**Comment by Lestropie** [comment]
Okay. In that case, I'd probably just accept the fact that the script will issue a warning if the CUDA version of `eddy` is not executed successfully, and remove the numba dependency.

In fact in retrospect this probably only happens if executed with heightened terminal verbosity: otherwise, the warning of `eddy_cuda` failing will be generated inside `dwipreproc`, and therefore never get escalated to the terminal. Indeed this will be improved later in the updated *MRtrix3* Python API (MRtrix3/mrtrix3#1449), which utilises custom exceptions, allowing me to catch and handle such things more sensibly.

**Comment by araikes** [comment]
Ok. So conceptually, just install the CUDA libs and let the script figure out if it can run CUDA.

Would you then not remove `eddy_cuda` in the build? If so, how would dwipreproc handle the `--mporder` flag if it tries to run CUDA and can't (assuming we can find slice timings as anticipated)?

**Comment by Lestropie** [comment]
The `--mporder` issue is an annoying one. The command help pages are interrogated to find out which command-line options they offer. Unfortunately the OpenMP version provides this option, but it immediately terminates the command if used. So the options presented in the help page can'e be 100% trusted. This is however again something that would need to be managed within `dwipreproc` rather than this script; that's really where all the automagic processing happens, this script just utilises it.

**Comment by araikes** [comment]
> This is however again something that would need to be managed within dwipreproc rather than this script; that's really where all the automagic processing happens, this script just utilises it.

I suppose perhaps this is where my confusion comes in about the interface between the present script and the `dwipreproc` is _supposed_ to function. Should `dwipreproc` take all available arguments and intelligently decide which ones are valid/run-able? It seem like no, since the script is parsing the help pages to decide which options are available and then feeding them to `dwipreproc` for execution.

Just so there isn't any confusion, the edits I made using `numba` weren't to alter the function of `dwipreproc` but rather argument parsing. I probably shouldn't have posted on my phone so I could actually paste in the code block, but here it is (added both in `ParticipantShared` and in the main function (with `self.` removed)):

```
        self.eddy_binary = fsl.eddyBinary(True)
        if self.eddy_binary:
                # Check to see if CUDA-enabled cores are available.
                try:
                        len(cuda.gpus)
                # Revert to eddy_openmp on error
                except:
                        self.eddy_binary = fsl.eddyBinary(False)
                        self.eddy_cuda = False
                        app.warn('eddy_cuda was found. However, no GPUs were available.')
                # Otherwise, CUDA!
                else:
                        self.eddy_cuda = True
        else:
            self.eddy_binary = fsl.eddyBinary(False)
            self.eddy_cuda = False
```

**Comment by Lestropie** [comment]
> I suppose perhaps this is where my confusion comes in about the interface between the present script and the way `dwipreproc` is *supposed* to function.

Not sure if there's any "supposed to" in there; it's however I choose to make it. But in general, the number of users of `dwipreproc` far exceeds the number of users of this script; so if there's an issue that requires resolution, and it could theoretically be handled in either location, `dwipreproc` makes more sense. For instance, with the latest version of `eddy` no longer producing the CNR / residual map images unless command-line options are specified explicitly, `dwipreproc -eddyqc_all` will no longer behave as it used to, even for *MRtrix3* users unaware of the existence of this tool. So:

> ... since the script is parsing the help pages to decide which options are available and then feeding them to `dwipreproc` for execution.

Given this observation, it's likely that `dwipreproc` will itself *also* have to query the `eddy` help page(s), so that it can determine which command-line options to pass to `eddy` in order to retain predictable behaviour.

It's probably the same for determining whether or not the CUDA version can be run, as well as any potential changes to command-line options necessary to get one or the other to work; these are decisions that need to be made for `dwipreproc` usage external to this tool also, so given I have the capability to modify both here and upstream, upstream probably makes more sense.

**Comment by araikes** [comment]
Well in that case I await the solution from the upstream changes :)

In the mean time, I'm running the current dev version of the script on my longitudinal dataset and it did a great job of parsing the data and appears to be doing everything as expected.

**Comment by Lestropie** [comment]
Should (finally) be fixed by upstream changes in #62.
